logstreams/src/main/java/io/zeebe/logstreams/state/StateSnapshotController.java

```diff
@@ -10,124 +10,84 @@
-      final StateStorage storage,
-      final SnapshotReplication replication,
-      final int maxSnapshotCount) {
+      final SnapshotStorage storage,
+      final SnapshotReplication replication) {
     this.storage = storage;
     this.zeebeDbFactory = zeebeDbFactory;
-    this.maxSnapshotCount = maxSnapshotCount;
-    this.replicationController = new ReplicationController(replication, storage, this);
-
-    initializeRestoreInfo();
+    this.replicationController = new ReplicationController(replication, storage);
   }
 
   @Override
-  public void takeSnapshot(final long lowerBoundSnapshotPosition) {
-    if (db == null) {
-      throw new IllegalStateException("Cannot create snapshot of not open database.");
-    }
+  public Snapshot takeSnapshot(final long lowerBoundSnapshotPosition) {
+    final var snapshot = storage.getPendingSnapshotFor(lowerBoundSnapshotPosition);
+    createSnapshot(snapshot.getPath());
+    storage.commitSnapshot(snapshot);
 
-    final File snapshotDir = storage.getSnapshotDirectoryFor(lowerBoundSnapshotPosition);
-    db.createSnapshot(snapshotDir);
-    snapshotInfo =
-        new DefaultSnapshotInfo(lowerBoundSnapshotPosition, snapshotDir.listFiles().length);
+    return snapshot;
   }
 
   @Override
-  public void takeTempSnapshot() {
-    if (db == null) {
-      throw new IllegalStateException("Cannot create snapshot of not open database.");
-    }
-
-    final File snapshotDir = storage.getTempSnapshotDirectory();
-    LOG.debug("Take temporary snapshot and write into {}.", snapshotDir.getAbsolutePath());
-    db.createSnapshot(snapshotDir);
+  public Snapshot takeTempSnapshot(final long lowerBoundSnapshotPosition) {
+    final var snapshot = storage.getPendingSnapshotFor(lowerBoundSnapshotPosition);
+    createSnapshot(snapshot.getPath());
+    return snapshot;
   }
 
   @Override
-  public void moveValidSnapshot(final long lowerBoundSnapshotPosition) throws IOException {
-    if (db == null) {
-      throw new IllegalStateException("Cannot create snapshot of not open database.");
-    }
-
-    final File previousLocation = storage.getTempSnapshotDirectory();
-    if (!previousLocation.exists()) {
-      throw new IllegalStateException(
-          String.format(
-              "Temporary snapshot directory %s does not exist.",
-              previousLocation.getAbsolutePath()));
-    }
-
-    final File snapshotDir = storage.getSnapshotDirectoryFor(lowerBoundSnapshotPosition);
-    if (snapshotDir.exists()) {
-      return;
-    }
-
-    LOG.debug(
-        "Snapshot is valid. Move snapshot from {} to {}.",
-        previousLocation.getAbsolutePath(),
-        snapshotDir.getAbsolutePath());
-
-    Files.move(previousLocation.toPath(), snapshotDir.toPath());
-    onNewValidSnapshot();
+  public void commitSnapshot(final Snapshot snapshot) throws IOException {
+    // TODO: why this check? it was previously here
+    Objects.requireNonNull(db, "Cannot commit snapshot for a closed database");
```

Comment:
Because it could be closed in between :man_shrugging: probably because we had some dependency issues before ^^

Reply:
But committing the snapshot should be ok even if the DB is not opened right? The snapshot was already taken

Reply:
then remove it